_posts/2022-11-12-Aya-gsoc-final-report.md

```diff
@@ -0,0 +1,132 @@
+
+# 1- Removing types and mass guessing from Topology Parsers ([commit 0cbc497]) 
+First, I navigated through all parsers to see where and how types and mass guessing take place. The below table shows the behavior of types and masses guessing inside parsers.
+![Guessing inside Parsers](public/images/final_report_aya/parsers_guessing.png)
+As we see, not all parsers use the same attribute in guessing masses. The DefaultGuesser guesses masses by first looking for the elements attribute and if not available, it looks for the types attribute, this behavior preserves the default behavior of all parsers except the ones that guess masses from names. Parsers that guess masses from types are [FHIAIMSParser], [TXYZParser], and [XYZParser]. For both FHIAIMSParser and XYZParser `element` attributes are provided, and luckily it is just a copy of the `names` attribute, so the default behavior of these parsers is not affected. For TXYZParser, I added a feature of publishing Elements attribute to its topology output in case all read names are valid elements ([merged PR #3836]).
+
+2. Developing the `guess_topologyAttributes` API
+As described in issue [# 3704], the new guesser API will have the tasks of guessing and adding new topology attributes to the [Universe]. The user only passes the context (eg. `default`, `pdb`, `martini`) and the attributes of interest to be guessed, either through the `to_guess` or `force_guess` parameters and the API does the rest. Since I was writing my proposal, I was keen to develop the guesser API with this high level of abstraction and make guessing attributes easy and straightforward to the user without bothering about which method should be called and which attribute(s) is used as a reference in guessing other attributes, so all this work is handled inside the API and the Guesser class. The `guess_topologyAttributes` make the following processes:
+
+- gets the appropriate Guesser class that matches the passed context.
+
+- checks if the Guesser class support guessing the attributes passed to the`to_guess` and/or `force_guess` parameters.
+
+- check if any attribute passed to the `to_guess` parameter already exists in the topology attributes of the [Universe]. If so, warn the user that only empty values will be filled for this attribute, if any exists and in case the user wishes to override all the attribute values, he must pass it to the `force_guess` parameter instead of the `to_guess` one.
+
+- guessing attributes is handled by the guess_attrs method, which is declared inside the parent guesser class GuesserBase. It manages partial and complete guessing of attributes and calls the appropriate guesser method for each attribute.
```

Comment:
`guess_attrs()` method

---

(monospaced and I add () after methods/functions)